lib/ace/transport_app.rb

```diff
@@ -33,6 +34,57 @@ def initialize(config = nil)
       super(nil)
     end
 
+    # Initialises the puppet target.
+    # @param certname   The certificate name of the target.
+    # @param transport  The transport provider of the target.
+    # @param target     Target connection hash or legacy connection URI
+    # @return [Puppet device instance] Returns Puppet device instance, or nil if exception thrown
+    # @raise  [puppetlabs/ace/invalid_param] If nil parameter or no connection detail found
+    # @example Connect to device.
+    #   init_puppet_target('test_device.domain.com', 'panos',  JSON.parse("target":{
+    #                                                                       "remote-transport":"panos",
+    #                                                                       "host":"fw.example.net",
+    #                                                                       "user":"foo",
+    #                                                                       "password":"wibble"
+    #                                                                     }) ) => panos.device
+    def self.init_puppet_target(certname, transport, target)
+      unless target
+        ACE::Error.new("There was an error parsing the Puppet target. 'target' not found",
+                       'puppetlabs/ace/invalid_param')
+        return nil
+      end
+      unless certname
+        ACE::Error.new("There was an error parsing the Puppet compiler details. 'certname' not found",
+                       'puppetlabs/ace/invalid_param')
+        return nil
+      end
+      unless transport
+        ACE::Error.new("There was an error parsing the Puppet target. 'transport' not found",
+                       'puppetlabs/ace/invalid_param')
+        return nil
+      end
+
+      if target['uri']
+        if target['uri'] =~ URI::DEFAULT_PARSER.make_regexp
+          # Correct URL
+          url = target['uri']
+        else
+          ACE::Error.new("There was an error parsing the URI of the Puppet target",
+                         'puppetlabs/ace/invalid_param')
+          return nil
+        end
+      else
+        url = Hash[target.map { |(k, v)| [k.to_sym, v] }]
+      end
+
+      device_struct = Struct.new(:provider, :url, :name, :options)
+      device = Puppet::Util::NetworkDevice.init(device_struct.new(transport,
+                                                                  url,
+                                                                  certname,
+                                                                  {}))
+      device.current
```

Comment:
Some findings `device.current` is not suited as a return, and simply `device` is sufficient.